### PR TITLE
Fix #1943: improve error message when an argument fails typing

### LINKED ISSUE #1943: nopython mode and endianness
If I try to pass some numba function an array with "endianess" dtype it crashes:

```
import numba as nb
import numpy as np

@nb.njit
def testfunc(arr):
    return arr

testfunc(np.ones((3, 3), dtype='>i8'))
TypeError: Failed at nopython (nopython frontend)
Arg 0 of pyobject is not legal in nopython mode
```

I don't know if it's planned to support these dtypes. If not, could unsupported dtypes raise a custom exception message? 

It's really hard to figure out such crashs if one doesn't know it's caused by the dtype.

### PR DESCRIPTION
**Pull request by pitrou**

### CODE REVIEW

File: numba/dispatcher.py [hunk 1]
```diff
@@ -353,9 +377,13 @@ def typeof_pyval(self, val):
         """
         # Not going through the resolve_argument_type() indirection
         # can shape a couple µs.
-        tp = typeof(val)
-        if tp is None:
+        try:
```
Comment (stuartarchibald) [inline numba/dispatcher.py]: Above comment https://github.com/numba/numba/pull/1948/files#diff-66adccf97cb99c68d3c33d8025531cc6R379 should probably say `save` not `shape`?!

Comment (pitrou) [inline numba/dispatcher.py]: Actually, I meant _shave_, but yes, too :-)

File: numba/tests/test_typeof.py [hunk 2]
```diff
@@ -108,6 +108,13 @@ def check(arr, ndim, layout, mutable, aligned):
         a4.flags.writeable = False
         check(a4, 0, 'C', False, True)
 
+        # Unsupported dtype
+        a5 = a1.astype(a1.dtype.newbyteorder())
+        with self.assertRaises(ValueError) as raises:
```
Comment (stuartarchibald) [inline numba/tests/test_typeof.py]: Would `assertRaisesRegex(p)` do the same?

**Comment by stuartarchibald** [comment]
Thanks for the patch, these new diagnostic messages are a good improvement.

**Pull request merged**